Don't specify default len/cap sizes to make

bradleyfalzon opened the pull request:
See testdata for the main tests, but essentially, for any type T:

- `make(T, 0)` can be simplified to `make(T)` for non slice types
- `make(T, 1, 1)` can be simplified to `make(T, 1)`

Notes:
- I'm ignoring the type, as whether the type is a slice, chan or map, the same rules can be applied
- I'm not convinced on the README wording
- `constantInt` is a function copied from `staticcheck`, if this issue is accepted, perhaps it might be better to move this function to `lintuils`?
- I can remove the redundant selector tests, I was just ensuring it was catching the sel.sel.ident case
- Is there a better way to compare idents or selectors? I feel I maybe doing more work than required.
- I haven't ran this through my smaller corpus, I will be doing over the next 48 hours (slowly updating it now), so I haven't checked for false positives, panics etc

bradleyfalzon committed:
Don't specify default len/cap sizes to make

dominikh commented:
I think we should check for the literal `0` here, not for any constant representing `0`. Those constants might be due to debugging, math or platform-specific code. The literal `0`, however, is pretty much always redundant.

> Is there a better way to compare idents or selectors? I feel I maybe doing more work than required.

Whether two expressions are the same can easily be tested with `f.Render(sel1) == f.Render(sel2)` -- there will be no ambiguity here due to the scopes involved (that is, there is no way any ident involved will change its meaning during the same call to make)

bradleyfalzon committed:
Simplify default len/cap size checks for make

bradleyfalzon commented:
OK, revised based on my understanding, using `f.Render` and removed extra tests. Also changed the README as well to be more in line with existing check descriptions and suggestions.

dmitshur commented:
> Whether two expressions are the same can easily be tested with `f.Render(sel1) == f.Render(sel2)` -- there will be no ambiguity here due to the scopes involved (that is, there is no way any ident involved will change its meaning during the same call to make)

What's the status of this case?

```Go
sideeffect := func() int {
    global++
    return 1
}

make(T, sideeffect(), sideeffect())
```

dominikh commented:
Theoretical potential for a false positive, practically speaking not a concern. These functions are either pure, or very poorly written code we're fine with having a false positive for.

dmitshur commented:
Hmm. I'm fine with that. Until I run into a false positive case which isn't poorly written code, then I'll raise an issue. I expect that to never happen.

dominikh commented:
`make(T)` is in fact not valid for slices, and the size is necessary.

bradleyfalzon commented:
That is a compiler error though, did you want additional checks?

dominikh commented:
Your current check, when it sees `make([]T, 0)`, will say that the length is redundant, suggesting `make([]T)` instead. That's an invalid suggestion. It shouldn't make that suggestion for slices.

bradleyfalzon committed:
Length is mandatory for slices, don't suggest removal

bradleyfalzon commented:
Yes, I see, thanks.

I've rebased and added another condition to skip checking length argument when the type argument  to make looks like a slice.

This won't report any suggestions when it finds `make([]T, 0)`, this is in line with https://github.com/golang/lint/issues/234.

dmitshur commented:
What about `type underlyingSlice []int; make(underlyingSlice, 0)` case?

bradleyfalzon commented:
Yeah thanks, I'll check and update.

bradleyfalzon committed:
Use types to check the underlying type provided to make

bradleyfalzon commented:
See https://github.com/dominikh/go-tools/pull/3
